Some documentation updates (#7328)
cython/cython: docs/src/userguide/wrapping_CPlusPlus.rst

    cdef int raise_py_error()
    cdef int something_dangerous() except +raise_py_error

If ``something_dangerous`` raises a C++ exception then ``raise_py_error`` will be
called during exception handling (i.e. typically, in a ``catch`` clause).
From there, ``raise_py_error`` should access the exception object by,
for example, rethrow the exception currently handled with ``throw;`` and
immediately catch it, or use ``std::current_exception()``; then, it should set
a Python exception, such as by ``PyErr_SetString``.
This allows one to do custom C++ to Python error "translations".
See the example below. If ``raise_py_error`` does not actually raise an exception a
``RuntimeError`` will be raised. This approach may also be used to manage custom

> Python exceptions created using the Python C API. ::

Reviewer: ```suggestion
If ``something_dangerous`` raises a C++ exception then ``raise_py_error`` will be
called during exception handling (i.e. typically, in a ``catch`` clause).
At that point, ``raise_py_error`` should retrieve the current exception object,
either by rethrowing the current exception with ``throw;`` and immediately catching it,
or via ``std::current_exception()``.  It can then set a Python exception via the ``PyErr_Set*()``
functions of the Python C-API to implement a custom error translation from C++ to Python.
See the example below. If  ``raise_py_error`` does not actually raise an exception, a
``RuntimeError`` will be raised.  This approach may also be used to manage custom
Python exceptions created using the Python C API. ::
```

or via ``std::current_exception()``.  It can then set a Python exception via the ``PyErr_Set*()``
functions of the Python C-API to implement a custom error translation from C++ to Python.
See the example below. If  ``raise_py_error`` does not actually raise an exception, a
``RuntimeError`` will be raised.  This approach may also be used to manage custom
Python exceptions created using the Python C API. ::

    from cpython.object cimport PyObject

    cdef extern from *:
        """
        #include <Python.h>
        #include <stdexcept>
        #include <ios>

        void cpp_handle_exception(PyObject *exc_type) {
            if (PyErr_Occurred())
                return; // let the latest Python exn pass through and ignore the current one
            try {
                throw;
            }  catch (const std::logic_error& exn) {
                PyErr_SetString(exc_type, exn.what());
            } catch (...) {
                PyErr_SetString(PyExc_RuntimeError, "Unknown exception");
            }
        }
		"""

        cdef void cpp_handle_exception(object)

    class CustomLogicError(Exception): pass

>     cdef void translate_cpp_exception() except*:

Reviewer: ```suggestion
    cdef void translate_cpp_exception() except *:
        # Pass our custom Python exception types into the C++ handler,
        # in case it wants to raise them.
```

found `here
<https://github.com/cython/cython/blob/master/Cython/Utility/CppSupport.cpp>`__.

You can reuse Cython's built-in logic with something like the following.
Notice the ``convert_current_exception_to_python()`` function. ::

    from cpython.object cimport PyObject

    cdef extern from *:
        """
        #include <Python.h>
        #include <stdexcept>
        #include <ios>

        void rethrow_current_exception() {
            throw;
        }

        int convert_current_exception_to_python();

        void cpp_handle_exception(PyObject *exc_type) {
            if (PyErr_Occurred())
                return;
            try {
                throw;
            }  catch (const std::logic_error& exn) {
                PyErr_SetString(exc_type, exn.what());
            } catch (...) {
                convert_current_exception_to_python();
            }
        }
        """

        cdef void rethrow_current_exception() except+

        cdef void cpp_handle_exception(object)

    cdef public int convert_current_exception_to_python() except -1:

>         rethrow_current_exception()  # the heavy lifting is done in the ``except+``

Reviewer: I think this can be a `void` function. It doesn't need Cython error propagation since it's not meant to be called from Cython. I'd also change the name a little to make the distinction between C++ exceptions and Python exceptions clearer.
```suggestion
    cdef public void convert_current_cpp_exception_to_python() except *:
        rethrow_current_cpp_exception()  # declared as ``except+``, which translates the exception
```

found `here
<https://github.com/cython/cython/blob/master/Cython/Utility/CppSupport.cpp>`__.

> You can reuse Cython's built-in logic with something like the following.

Reviewer: ```suggestion
You can reuse Cython's own C++ exception mapping with something like the following.
```

found `here
<https://github.com/cython/cython/blob/master/Cython/Utility/CppSupport.cpp>`__.

You can reuse Cython's built-in logic with something like the following.
Notice the ``convert_current_exception_to_python()`` function. ::

    from cpython.object cimport PyObject

    cdef extern from *:
        """
        #include <Python.h>
        #include <stdexcept>
        #include <ios>

        void rethrow_current_exception() {
            throw;
        }

        PyObject *CustomLogicError;

        void create_custom_exceptions() {
            CustomLogicError = PyErr_NewException("raiser.CustomLogicError", NULL, NULL);
        }

>         extern int convert_current_exception_to_python();

Reviewer: I don't think the `extern` is needed, is it?
```suggestion
        int convert_current_exception_to_python();
```
